## **Importación de librerías**

In [25]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly as px
import seaborn as sns


# **1. Extracción**📌

## &nbsp;&nbsp;&nbsp;1.1. **Extracción de datos**

### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.1.1. **Cargando los datos de la API**

In [26]:
path = 'https://raw.githubusercontent.com/ingridcristh/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json'
df = pd.read_json(path)

In [27]:
df.sample(10)

customerID Churn                                           customer  \
5800  7929-SKFGK    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
665   0946-CLJTI   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
765   1090-ESELR    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3214  4486-EFAEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2100  2956-GGUCQ   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
788   1114-CENIM   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2330  3249-VHRIP    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
2195  3070-FNFZQ    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
131   0220-EBGCE   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
4808  6584-VQMYT    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
5800    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
665     {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
765      {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3214     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
2100    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
788     {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2330  {'PhoneService': 'No', 'MultipleLines': 'No ph...   
2195     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
131      {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4808     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                               internet  \
5800  {'InternetService': 'Fiber optic', 'OnlineSecu...   
665   {'InternetService': 'Fiber optic', 'OnlineSecu...   
765   {'InternetService': 'Fiber optic', 'OnlineSecu...   
3214  {'InternetService': 'No', 'OnlineSecurity': 'N...   
2100  {'InternetService': 'Fiber optic', 'OnlineSecu...   
788   {'InternetService': 'DSL', 'OnlineSecurity': '...   
2330  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2195  {'InternetService': 'No', 'OnlineSecurity': 'N...   
131   {'InternetService': 'Fiber optic', 'OnlineSecu...   
4808  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
5800  {'Contract': 'Two year', 'PaperlessBilling': '...  
665   {'Contract': 'Month-to-month', 'PaperlessBilli...  
765   {'Contract': 'Two year', 'PaperlessBilling': '...  
3214  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2100  {'Contract': 'Month-to-month', 'PaperlessBilli...  
788   {'Contract': 'Month-to-month', 'PaperlessBilli...  
2330  {'Contract': 'Two year', 'PaperlessBilling': '...  
2195  {'Contract': 'Two year', 'PaperlessBilling': '...  
131   {'Contract': 'Month-to-month', 'PaperlessBilli...  
4808  {'Contract': 'One year', 'PaperlessBilling': '...

* ### El DataFrame posee 6 columnas de las cuales 2 estan planas y 4 estan por normlaizar

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


* ### Normalizando columnas

In [29]:
#Normalizamos las columnas de uno en uno
customer_df = pd.json_normalize(df['customer'])
phone_df = pd.json_normalize(df['phone'])
internet_df = pd.json_normalize(df['internet'])
account_df = pd.json_normalize(df['account'])

In [30]:
#Creamos un data frame con las columnas planas
columnas_planas = df.drop(columns=['customer','phone','internet','account'])

### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.1.2. **Convirtiendo los datos en un DataFrame**

In [31]:
# Unimos todos los datos en un solo dataframe consolidado llamado "Datos"
datos = pd.concat([columnas_planas, customer_df, phone_df, internet_df, account_df], axis=1)

In [32]:
datos.sample(10)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
6157  8429-XIBUM   Yes    Male              0      No         No      22   
672   0956-ACVZC    No  Female              0      No         No      54   
953   1343-EHPYB   Yes    Male              0     Yes         No      63   
5785  7905-TVXTA    No  Female              0      No         No       5   
5546  7596-IIWYC    No  Female              0      No         No      27   
3713  5125-CNDSP    No    Male              0      No        Yes      63   
2110  2962-XPMCQ   Yes    Male              0      No         No       2   
5107  6980-IMXXE    No  Female              0     Yes        Yes      69   
5708  7799-LGRDP    No  Female              0      No         No      43   
844   1197-BVMVG    No  Female              1      No         No       4   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
6157          Yes               Yes     Fiber optic  ...                   No   
672           Yes                No     Fiber optic  ...                   No   
953           Yes               Yes     Fiber optic  ...                  Yes   
5785          Yes                No             DSL  ...                   No   
5546          Yes                No              No  ...  No internet service   
3713           No  No phone service             DSL  ...                  Yes   
2110          Yes                No             DSL  ...                   No   
5107          Yes                No              No  ...  No internet service   
5708          Yes               Yes              No  ...  No internet service   
844           Yes               Yes     Fiber optic  ...                   No   

         DeviceProtection          TechSupport          StreamingTV  \
6157                   No                   No                  Yes   
672                   Yes                   No                  Yes   
953                   Yes                   No                  Yes   
5785                   No                  Yes                  Yes   
5546  No internet service  No internet service  No internet service   
3713                   No                  Yes                   No   
2110                   No                   No                   No   
5107  No internet service  No internet service  No internet service   
5708  No internet service  No internet service  No internet service   
844                    No                   No                   No   

          StreamingMovies        Contract PaperlessBilling  \
6157                  Yes  Month-to-month              Yes   
672                   Yes        One year              Yes   
953                   Yes  Month-to-month              Yes   
5785                   No  Month-to-month              Yes   
5546  No internet service        Two year              Yes   
3713                   No        One year              Yes   
2110                   No  Month-to-month              Yes   
5107  No internet service        Two year               No   
5708  No internet service        Two year              Yes   
844                    No  Month-to-month              Yes   

                  PaymentMethod Charges.Monthly  Charges.Total  
6157  Bank transfer (automatic)          101.35         2317.1  
672     Credit card (automatic)           94.10         5060.9  
953            Electronic check          103.40           6603  
5785           Electronic check           64.35         352.65  
5546  Bank transfer (automatic)           20.25          538.2  
3713  Bank transfer (automatic)           40.60        2588.95  
2110           Electronic check           45.15           98.5  
5107  Bank transfer (automatic)           20.20        1412.65  
5708    Credit card (automatic)           25.70         1188.2  
844            Electronic check           74.45         294.45  

[10 rows x 21 columns]

# **2. Transformación**🔧

## &nbsp;&nbsp;&nbsp;2.1. **Conociendo el conjunto de datos**

### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.1.1. **Explorando las columnas del dataset y tipos de datos**

In [33]:
# Verificando las columnas y tipos de datos
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [34]:
# Verificando el tipo de dato de cada columna
datos.dtypes

customerID           object
Churn                object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
Charges.Monthly     float64
Charges.Total        object
dtype: object

### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.2.2. **Identificar las columnas más relevantes para el análisis**

* variable objetivo o target en este caso sería la columna "**churn**" ya que indica si el cliente abandonó el servicio.

* Variables categóricas

    * **CustomerID**: es el código de identificación del cliente
    * **gender**: género masculino o femenino
    * **Partner**: indica si el cleinte tiene o no tiene pareja (si / no)
    * **Dependents**: si el cliente tiene dependientes (si / no)
    * **PhoneService**: si esta o no esta suscrito al servicio telefonico (si / no)
    * **MultipleLines**: si esta suscrito a más de una linea telefónia (si / no)
    * **InternetService**: si esta  no esta a un proveedor de internet (si / no)
    * **OnlineSecurity**: suscripcion al servicio de seguridad en linea (si / no)
    * **OnlieBackup**: suscripción a un almacenamiento en linea (si / no)
    * **DeviceProtection**: Servicio de protección en linea (si / no)
    * **TechSupport**: servicio de soporte tecnico (si / no)
    * **StreamingTV**: Suscripción a televisión por cable (si / no)
    * **StreamingMovies**: Suscripción a peliculas de streaming (si / no)
    * **Contract**: indica el tiempo de contrato
    * **PaperlessBilling**: facturación en linea (si / no)
    * **PaymentMethod**: indica el metodo de pago del cliente

* Variable numericas

    * **tenure**: número de meses de contrato del cliente
    * **Charges.Monthly**: número total de servicios del cliente por mes
    * **Charges.Total**: gasto total del cliente

## &nbsp;&nbsp;&nbsp;2.2. **Comprobación de incoherencias en los datos**

### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.2.1. **Verificación de posibles problemas en los datos**

In [35]:
# Verificando si las columnas presentan valores nulos
datos.isnull().sum()

customerID          0
Churn               0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges.Monthly     0
Charges.Total       0
dtype: int64

In [36]:
# Verificando si los datos presentan valores duplicados

datos.duplicated().sum()

0

### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.2.2. **Manejo de inconsistencias**

In [39]:
# Eliminando filas en las que lacolumna 'Charges.Total' contega valores nulos

datos = datos.dropna(subset=['Charges.Total'])
datos

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60          593.3  
1                  No      Mailed check           59.90          542.4  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90          267.4  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15          742.9  
7263              Yes  Electronic check           85.10         1873.7  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00         3707.6  

[7267 rows x 21 columns]

In [65]:
datos['Charges.Total'].sample(10)

4264     2893.4
7014     1617.5
5223    1345.65
2021     3470.8
5550     4512.7
7248     367.55
6682     1777.9
4885     4448.8
6714       45.6
704      492.55
Name: Charges.Total, dtype: object

In [66]:
# Reemplazando posibles caracteres no numericos para preparar la columna 'Caharges.total' para su conversión

datos['Charges.Total'] = datos['Charges.Total'].str.replace(' ','').replace('$', '').replace(',', '').replace('', None)

In [67]:
datos['Charges.Total'].dtype

dtype('O')

In [68]:
# Cambiando tipo de dato de 'Charges.Total' de objet a float64

datos['Charges.Total'] = datos['Charges.Total'].astype(float)

In [69]:
datos['Charges.Total'].dtype

dtype('float64')

In [70]:
# Verificamos si la columna 'churn' tiene campos vacíos, realizando un conteo de valores únicos

datos['Churn'].value_counts()

Churn
No     5174
Yes    1869
        224
Name: count, dtype: int64

Se evidencia que hay 224 filas con valores en blanco

In [72]:
# Verificando filas con valores de 'churn' en blanco

datos[datos['Churn'].isin(['', None])]

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
30    0047-ZHDTW        Female              0      No         No      11   
75    0120-YZLQA          Male              0      No         No      71   
96    0154-QYHJU          Male              0      No         No      29   
98    0162-RZGMZ        Female              1      No         No       5   
175   0274-VVQOQ          Male              1     Yes         No      65   
...          ...   ...     ...            ...     ...        ...     ...   
7158  9840-GSRFX        Female              0      No         No      14   
7180  9872-RZQQB        Female              0     Yes         No      49   
7211  9920-GNDMB          Male              0      No         No       9   
7239  9955-RVWSC        Female              0     Yes        Yes      67   
7247  9966-VYRTZ        Female              0     Yes        Yes      31   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
30            Yes               Yes     Fiber optic  ...                   No   
75            Yes                No              No  ...  No internet service   
96            Yes                No             DSL  ...                  Yes   
98            Yes                No             DSL  ...                  Yes   
175           Yes               Yes     Fiber optic  ...                  Yes   
...           ...               ...             ...  ...                  ...   
7158          Yes               Yes             DSL  ...                  Yes   
7180           No  No phone service             DSL  ...                   No   
7211          Yes               Yes     Fiber optic  ...                   No   
7239          Yes                No              No  ...  No internet service   
7247          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
30                     No                   No                   No   
75    No internet service  No internet service  No internet service   
96                     No                  Yes                   No   
98                     No                  Yes                   No   
175                   Yes                   No                  Yes   
...                   ...                  ...                  ...   
7158                   No                   No                   No   
7180                   No                   No                  Yes   
7211                   No                   No                   No   
7239  No internet service  No internet service  No internet service   
7247  No internet service  No internet service  No internet service   

          StreamingMovies        Contract PaperlessBilling  \
30                     No  Month-to-month              Yes   
75    No internet service        Two year              Yes   
96                     No        One year              Yes   
98                     No  Month-to-month               No   
175                   Yes        One year              Yes   
...                   ...             ...              ...   
7158                   No        One year              Yes   
7180                   No  Month-to-month               No   
7211                   No  Month-to-month              Yes   
7239  No internet service        Two year              Yes   
7247  No internet service  Month-to-month              Yes   

                  PaymentMethod Charges.Monthly  Charges.Total  
30    Bank transfer (automatic)           79.00         929.30  
75      Credit card (automatic)           19.90        1355.10  
96             Electronic check           58.75        1696.20  
98      Credit card (automatic)           59.90         287.85  
175   Bank transfer (automatic)          103.15        6792.45  
...                         ...             ...            ...  
7158               Mailed check           54.25         773.20  
71

In [73]:
# Creando un nuevo DataFrame para excluir los valores nulos de la columna 'churn'

datos_analisis = datos[datos['Churn'] != '']



In [75]:
datos_analisis['Churn'].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [76]:
datos_analisis

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60         593.30  
1                  No      Mailed check           59.90         542.40  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90         267.40  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15         742.90  
7263              Yes  Electronic check           85.10        1873.70  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00        3707.60  

[7043 rows x 21 columns]

#📊 Carga y análisis

#📄Informe final